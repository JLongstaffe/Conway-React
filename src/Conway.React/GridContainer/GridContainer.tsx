
import React, { FunctionComponent, useState } from 'react';

import useInterval from '../useInterval';

import Grid from './Grid/Grid';

const GridContainer: FunctionComponent<IGridProperties> =
    ({ initialGrid, nextState, updateFrequency }) =>
{
    const [grid, setGrid] = useState(initialGrid);

    const onNext = () => setGrid(nextState(grid));

    const [playing, setPlaying] = useInterval(onNext, updateFrequency);

    const togglePlay = () => setPlaying(!playing);

    return <>
             <Grid grid={grid} setGrid={setGrid} />

             <button onClick={onNext}>Next</button>
             <button onClick={togglePlay}>{playing ? 'Pause' : 'Play'}</button>
           </>;
}

interface IGridProperties
{
    initialGrid: Readonly2D<boolean>;

    nextState: (state: Readonly2D<boolean>) => Readonly2D<boolean>;

    updateFrequency: number;
}

type Readonly2D<T> = ReadonlyArray<ReadonlyArray<T>>;

export default GridContainer;
