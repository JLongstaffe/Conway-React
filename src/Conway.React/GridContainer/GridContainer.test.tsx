
import React from 'react'

import { render, fireEvent, act } from '@testing-library/react';

import GridContainer from './GridContainer';

import { IStateGenerator } from '../../Conway.Core/Conway';

test('Grid snapshots', () =>
{
    const horizontal = render
        (<GridContainer initialGrid={horizontalState}
                        nextState={_ => []}
                        updateFrequency={0} />)
        .container;

    expect(horizontal.firstChild).toMatchSnapshot();

    const vertical = render
        (<GridContainer initialGrid={verticalState}
                        nextState={_ => []}
                        updateFrequency={0} />)
        .container;

    expect(vertical.firstChild).toMatchSnapshot();
});

test('Clicking on cell toggles alive state', () =>
{
    const { getByTestId } = render
        (<GridContainer initialGrid={horizontalState}
                        nextState={_ => []}
                        updateFrequency={0} />);

    const cell = getByTestId('1-1');

    const aliveBackground = cell.style.backgroundColor;

    fireEvent.click(cell);

    let toggledBackground = cell.style.backgroundColor;

    expect(toggledBackground).not.toBe(aliveBackground);

    fireEvent.click(cell);

    toggledBackground = cell.style.backgroundColor;

    expect(toggledBackground).toBe(aliveBackground);
});

test('Clicking Next advances next Conway state', () =>
{
    const mockNextState: jest.MockedFunction<IStateGenerator> = jest.fn();

    mockNextState.mockReturnValue(verticalState);

    let { getByText, asFragment } = render
        (<GridContainer initialGrid={horizontalState}
                        nextState={mockNextState}
                        updateFrequency={0} />);

    const nextButton = getByText('Next');

    fireEvent.click(nextButton);

    expect(mockNextState).toHaveBeenCalledTimes(1);

    const secondRenderHtml = asFragment().firstElementChild?.outerHTML;

    expect(secondRenderHtml).toEqual(getGridHtml(verticalState));
});

test('Clicking Play button toggles button text', () =>
{
    const { queryByText } = render
        (<GridContainer initialGrid={horizontalState}
                        nextState={_ => []}
                        updateFrequency={0} />);

    const button = queryByText('Play') as HTMLButtonElement;

    fireEvent.click(button);

    expect(queryByText('Pause')).not.toBeNull();

    fireEvent.click(button);

    expect(queryByText('Play')).not.toBeNull();
});

test('Play updates state with correct frequency', () =>
{
    jest.useFakeTimers();

    const mockNextState: jest.MockedFunction<IStateGenerator> = jest.fn();

    mockNextState.mockReturnValue([]);

    const { getByText } = render
        (<GridContainer initialGrid={[]}
                        nextState={mockNextState}
                        updateFrequency={3} />);

    expect(mockNextState).not.toBeCalled();

    act(() =>
    {
        fireEvent.click(getByText('Play'));

        jest.advanceTimersByTime(16);
    });

    expect(mockNextState).toHaveBeenCalledTimes(5);
});

test('Clicking Pause stops periodic updates', () =>
{
    jest.useFakeTimers();

    const mockNextState: jest.MockedFunction<IStateGenerator> = jest.fn();

    mockNextState.mockReturnValue([]);

    const { getByText } = render
        (<GridContainer initialGrid={[]}
                        nextState={mockNextState}
                        updateFrequency={3} />);

    expect(mockNextState).not.toBeCalled();

    act(() =>
    {
        fireEvent.click(getByText('Play'));

        jest.advanceTimersByTime(10);

        fireEvent.click(getByText('Pause'));

        jest.advanceTimersByTime(6);
    });

    expect(mockNextState).toHaveBeenCalledTimes(3);
});

function getGridHtml(state: Readonly2D<boolean>)
{
    return render
        (<GridContainer initialGrid={state}
                        nextState={_ => []}
                        updateFrequency={0} />)
        .asFragment()
        .firstElementChild?.outerHTML;
}

const horizontalState = To2DBoolean([[ 0, 0, 0 ],
                                     [ 1, 1, 1 ],
                                     [ 0, 0, 0 ]]);

const verticalState = To2DBoolean([[ 0, 1, 0 ],
                                   [ 0, 1, 0 ],
                                   [ 0, 1, 0 ]]);

function To2DBoolean(input: number[][]): boolean[][]
{
    return input.map(x => x.map(y => y === 1));
}

type Readonly2D<T> = ReadonlyArray<ReadonlyArray<T>>;