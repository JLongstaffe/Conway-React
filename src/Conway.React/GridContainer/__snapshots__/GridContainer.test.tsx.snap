// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid snapshots 1`] = `
<table>
  <tbody>
    <tr>
      <td
        data-testid="0-0"
        style="background-color: white;"
      >
         
      </td>
      <td
        data-testid="1-0"
        style="background-color: white;"
      >
         
      </td>
      <td
        data-testid="2-0"
        style="background-color: white;"
      >
         
      </td>
    </tr>
    <tr>
      <td
        data-testid="0-1"
        style="background-color: grey;"
      >
         
      </td>
      <td
        data-testid="1-1"
        style="background-color: grey;"
      >
         
      </td>
      <td
        data-testid="2-1"
        style="background-color: grey;"
      >
         
      </td>
    </tr>
    <tr>
      <td
        data-testid="0-2"
        style="background-color: white;"
      >
         
      </td>
      <td
        data-testid="1-2"
        style="background-color: white;"
      >
         
      </td>
      <td
        data-testid="2-2"
        style="background-color: white;"
      >
         
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Grid snapshots 2`] = `
<table>
  <tbody>
    <tr>
      <td
        data-testid="0-0"
        style="background-color: white;"
      >
         
      </td>
      <td
        data-testid="1-0"
        style="background-color: grey;"
      >
         
      </td>
      <td
        data-testid="2-0"
        style="background-color: white;"
      >
         
      </td>
    </tr>
    <tr>
      <td
        data-testid="0-1"
        style="background-color: white;"
      >
         
      </td>
      <td
        data-testid="1-1"
        style="background-color: grey;"
      >
         
      </td>
      <td
        data-testid="2-1"
        style="background-color: white;"
      >
         
      </td>
    </tr>
    <tr>
      <td
        data-testid="0-2"
        style="background-color: white;"
      >
         
      </td>
      <td
        data-testid="1-2"
        style="background-color: grey;"
      >
         
      </td>
      <td
        data-testid="2-2"
        style="background-color: white;"
      >
         
      </td>
    </tr>
  </tbody>
</table>
`;
